File: HW/HW8/LongestPalindrome.java
package HW.HW8;
import java.util.HashMap;
import java.util.Set;
public class LongestPalindrome {
    public int longestPalindrome(String s) {
        int longest = 0;
        if (s.length()==0)
            return 0;
        HashMap<Character,Integer> map = new HashMap<>();
        for (int i=0;i<s.length();i++){
            if (map.containsKey(s.charAt(i)))
                map.put(s.charAt(i),map.get(s.charAt(i))+1);
            else
                map.put(s.charAt(i),1);
        }
        char maxchar = '!';
        for (int i=0;i<s.length();i++){
            if (map.get(s.charAt(i))%2==1 && (maxchar=='!'||map.get(s.charAt(i))>map.get(maxchar)))
                maxchar=s.charAt(i);
        }
        if (map.containsKey(maxchar)){
            longest=map.get(maxchar);
            map.remove(maxchar);
        }
        Set<Character> keys = map.keySet();
        for (Character c: keys){
            if (map.get(c)%2==1)
                longest+=map.get(c)-1;
            else
                longest+=map.get(c);
        }
        return longest;
    }
    public static void main(String[] args){
        LongestPalindrome a = new LongestPalindrome();
        System.out.println(a.longestPalindrome(""));
        System.out.println(a.longestPalindrome("abccccdd"));
        System.out.println(a.longestPalindrome("speediskey"));
        System.out.println(a.longestPalindrome("bb"));
    }
}

File: HW/HW8/LongestPalindrome.py
class LongestPalindrome:
    def longestPalindrome(self, s):
        if len(s)==0 or len(s)==1:
            return len(s)
        map = {}
        
        for char in s:
            if char in map:
                map[char]+=1
            else:
                map[char]=1
        maxchar = None
        for char in s:
            if map[char] % 2 == 1 and (maxchar is None or map[char] > map[maxchar]):
                maxchar = char
        
        if maxchar in map:
            longest = map[maxchar]
            map[maxchar]=0
        else:
            longest=0
        
        for char in map.keys():
            if map[char]%2==1:
                longest+=map[char]-1
            else:
                longest+=map[char]
        return longest

a = LongestPalindrome()
print(a.longestPalindrome(""))
print(a.longestPalindrome("abccccdd"))
print(a.longestPalindrome("speediskey"))
print(a.longestPalindrome("bb"))
            
